lib/libunw/unwind_arm32.c

```diff
@@ -0,0 +1,509 @@
+		DMSG("Unhandled instruction %.2x", insn);
+		return false;
+	}
+
+	if (update_vsp)
+		state->registers[SP] = vsp;
+
+	return true;
+}
+
+/* Performs the unwind of a function */
+static bool unwind_tab(struct unwind_state_arm32 *state, vaddr_t stack,
+		       size_t stack_size)
+{
+	uint32_t entry;
+	uint32_t insn;
+
+	/* Set PC to a known value */
+	state->registers[PC] = 0;
+
+	if (!copy_in(&insn, (void *)state->insn, sizeof(insn))) {
+		DMSG("Bad insn addr %p", (void *)state->insn);
+		return true;
+	}
+
+	/* Read the personality */
+	entry = insn & ENTRY_MASK;
+
+	if (entry == ENTRY_ARM_SU16) {
+		state->byte = 2;
+		state->entries = 1;
+	} else if (entry == ENTRY_ARM_LU16) {
+		state->byte = 1;
+		state->entries = ((insn >> 16) & 0xFF) + 1;
+	} else {
+		DMSG("Unknown entry: %x", entry);
+		return true;
+	}
+
+	while (state->entries > 0) {
+		if (!unwind_exec_insn(state, stack, stack_size))
+			return true;
+	}
+
+	/*
+	 * The program counter was not updated, load it from the link register.
+	 */
+	if (state->registers[PC] == 0) {
+		state->registers[PC] = state->registers[LR];
+
+		/*
+		 * If the program counter changed, flag it in the update mask.
+		 */
+		if (state->start_pc != state->registers[PC])
+			state->update_mask |= 1 << PC;
+	}
+
+	return false;
+}
+
+bool unwind_stack_arm32(struct unwind_state_arm32 *state,
+			vaddr_t stack, size_t stack_size)
+{
+	struct unwind_idx *index;
+	bool finished;
+
+	/* Reset the mask of updated registers */
+	state->update_mask = 0;
+
+	/* The pc value is correct and will be overwritten, save it */
+	state->start_pc = state->registers[PC];
+
+	/*
+	 * Find the item to run. Subtract 2 from PC to make sure that we're
+	 * still inside the calling function in case a __no_return function
+	 * (typically panic()) is called unconditionally and may cause LR and
+	 * thus this PC to point into the next and entirely unrelated function.
+	 */
+	index = find_index(state->start_pc - 2);
+	if (!index)
+		return false;
+
+	finished = false;
+	if (index->insn != EXIDX_CANTUNWIND) {
+		if (index->insn & (1U << 31)) {
+			/* The data is within the instruction */
+			state->insn = (vaddr_t)&index->insn;
+		} else {
+			/* A prel31 offset to the unwind table */
+			state->insn = (vaddr_t)&index->insn +
+				      expand_prel31(index->insn);
+		}
+
+		/* Run the unwind function */
+		finished = unwind_tab(state, stack, stack_size);
+	}
+
+	/* This is the top of the stack, finish */
+	if (index->insn == EXIDX_CANTUNWIND)
+		finished = true;
+
+	return !finished;
+}
+
+void print_stack_arm32(struct unwind_state_arm32 *state,
+		       vaddr_t stack, size_t stack_size)
+{
+	trace_printf_helper_raw(TRACE_ERROR, true, "Call stack:");
+	do {
+		trace_printf_helper_raw(TRACE_ERROR, true, " 0x%08" PRIx32,
+					state->registers[PC]);
+	} while (unwind_stack_arm32(state, stack, stack_size));
+}
+
+#if defined(__KERNEL__) && defined(ARM32)
+bool find_exidx(vaddr_t addr __unused, vaddr_t *idx_start, vaddr_t *idx_end)
```

Comment:
Wouldn't it make more sense to keep this outside libunw?
This depends on internals of Core. For ldelf the counterpart is actually in ldelf.
Same goes for for the other functions below, they seem more suited to keep in Core.